diff --git a/src/core/lib/security/credentials/tls/tls_credentials.cc b/src/core/lib/security/credentials/tls/tls_credentials.cc
index 701fd3b150..0826d05933 100644
--- a/src/core/lib/security/credentials/tls/tls_credentials.cc
+++ b/src/core/lib/security/credentials/tls/tls_credentials.cc
@@ -111,9 +111,9 @@ TlsServerCredentials::create_security_connector() {
 
 grpc_channel_credentials* grpc_tls_credentials_create(
     grpc_tls_credentials_options* options) {
-  if (!CredentialOptionSanityCheck(options, true /* is_client */)) {
-    return nullptr;
-  }
+  //if (!CredentialOptionSanityCheck(options, true /* is_client */)) {
+  //  return nullptr;
+  //}
   return new TlsCredentials(
       grpc_core::RefCountedPtr<grpc_tls_credentials_options>(options));
 }
